encoding/codecv7.go

```diff
@@ -0,0 +1,319 @@
+		block.Header.GasLimit,        // gasLimit
+		uint16(numTransactions),      // numTransactions
+		numL1Messages,                // numL1Messages
+	)
+
+	return daBlock, nil
+}
+
+// NewDAChunk creates a new DAChunk from the given Chunk and the total number of L1 messages popped before.
+// Note: For DACodecV7, this function is not implemented since there is no notion of DAChunk in this version. Blobs
+// contain the entire batch data without any information of Chunks within.
+func (d *DACodecV7) NewDAChunk(_ *Chunk, _ uint64) (DAChunk, error) {
+	return nil, nil
+}
+
+// NewDABatch creates a DABatch including blob from the provided Batch.
+func (d *DACodecV7) NewDABatch(batch *Batch) (DABatch, error) {
+	if len(batch.Chunks) != 0 {
+		return nil, errors.New("batch must not contain any chunks")
+	}
+
+	if len(batch.Blocks) == 0 {
+		return nil, errors.New("batch must contain at least one block")
+	}
+
+	blob, blobVersionedHash, blobBytes, err := d.constructBlob(batch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct blob: %w", err)
+	}
+
+	daBatch, err := newDABatchV7(CodecV7, batch.Index, blobVersionedHash, batch.ParentBatchHash, blob, blobBytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct DABatch: %w", err)
+	}
+
+	return daBatch, nil
+}
+
+func (d *DACodecV7) constructBlob(batch *Batch) (*kzg4844.Blob, common.Hash, []byte, error) {
+	enableCompression, err := d.CheckBatchCompressedDataCompatibility(batch)
+	if err != nil {
+		return nil, common.Hash{}, nil, fmt.Errorf("failed to check batch compressed data compatibility: %w", err)
+	}
+
+	blobBytes := make([]byte, blobEnvelopeV7OffsetPayload)
+	blobBytes[blobEnvelopeV7OffsetVersion] = uint8(CodecV7)
+
+	payloadBytes, err := d.constructBlobPayload(batch)
+	if err != nil {
+		return nil, common.Hash{}, nil, fmt.Errorf("failed to construct blob payload: %w", err)
+	}
+
+	if enableCompression {
+		// compressedPayloadBytes represents the compressed blob payload
+		var compressedPayloadBytes []byte
+		compressedPayloadBytes, err = zstd.CompressScrollBatchBytes(payloadBytes)
+		if err != nil {
+			return nil, common.Hash{}, nil, fmt.Errorf("failed to compress blob payload: %w", err)
+		}
+		// Check compressed data compatibility.
+		if err = checkCompressedDataCompatibility(compressedPayloadBytes); err != nil {
+			log.Error("ConstructBlob: compressed data compatibility check failed", "err", err, "payloadBytes", hex.EncodeToString(payloadBytes), "compressedPayloadBytes", hex.EncodeToString(compressedPayloadBytes))
+			return nil, common.Hash{}, nil, err
+		}
+		blobBytes[blobEnvelopeV7OffsetCompressedFlag] = 0x1
+		payloadBytes = compressedPayloadBytes
+	} else {
+		blobBytes[blobEnvelopeV7OffsetCompressedFlag] = 0x0
+	}
+
+	sizeSlice := encodeSize3Bytes(uint32(len(payloadBytes)))
+	copy(blobBytes[blobEnvelopeV7OffsetByteSize:blobEnvelopeV7OffsetCompressedFlag], sizeSlice)
+	blobBytes = append(blobBytes, payloadBytes...)
```

Comment:
Small nitpick that it's simpler to follow if you prepare the data first then encode in one code block, in encoding order

```go
blobBytes[blobEnvelopeV7OffsetVersion] = uint8(CodecV7)
copy(blobBytes[blobEnvelopeV7OffsetByteSize:blobEnvelopeV7OffsetCompressedFlag], sizeSlice)
blobBytes[blobEnvelopeV7OffsetCompressedFlag] = isCompressed
blobBytes = append(blobBytes, payloadBytes...)
```

Reply:
Done in cbed8b2507be70f82e6cb4663668c56ed0c72edf